project/src/main/java/com/google/sps/utilities/SpannerTasks.java

```diff
@@ -292,4 +306,82 @@ public static void updateVolunteeringOpportunity(VolunteeringOpportunity opportu
     }
     return results;
   }
+
+  private static List<Mutation> getMutationsFromBuilder(
+      Mutation.WriteBuilder builder, OpportunitySignup signup) {
+    List<Mutation> mutations = new ArrayList<>();
+    builder
+        .set(OPPORTUNITY_ID)
+        .to(signup.getOpportunityId())
+        .set(EMAIL)
+        .to(signup.getEmail());
+    mutations.add(builder.build());
+    return mutations;
+  }
+
+  /**
+   * Given a signup, insert a row with all available fields into the database
+   * and decrement the number of spots for the corresponding volunteering 
+   * opportunity in the database.
+   *
+   * @param signup the signup to be inserted
+   */
+  public static void insertOpportunitySignup(OpportunitySignup signup) {
+    SpannerClient.getDatabaseClient()
+        .readWriteTransaction()
+        .run(
+            new TransactionCallable<Void>() {
+              @Override
+              public Void run(TransactionContext transaction) throws Exception {
+                Struct row =
+                    transaction.readRow(
+                        VOLUNTEERING_OPPORTUNITY_TABLE,
+                        Key.of(signup.getOpportunityId()),
+                        Arrays.asList(NUM_SPOTS_LEFT));
+                long numSpotsLeft = row.getLong(0);
+                if (numSpotsLeft > 0) {
```

Comment:
is it possible for numSpotsLeft == 0 or < 0? (Is it possible for someone to accidentally sign up to volunteer for an event w no opportunities left?)